### PR TITLE
Settings derive macro

### PR DESCRIPTION
**Pull request by tim-harding**
I rewrote the settings registration to use a derive macro rather than `macro_rules!`. Using `#[derive(SettingGroup)]` will create a static method on the struct called `register` that serves the same role as calls to `register_nvim_setting` did previously. The setting names can be prefixed by adding a struct attribute of the form `#[setting_prefix = "my_prefix"]`. This is just another quality of life refactor, largely motivated by my interest in learning how the language feature works. 

For some reason, there are huge diffs in `renderer/mod.rs`, `bridge/mod.rs`, and `main.rs`. Looking at what changed, it appears likely that the recent upstream versions of those files didn't have `cargo fmt` run on them or something. I did not modify `bridge/mod.rs` and `renderer/mod.rs` only needs the changes related to `RendererSettings` early on in the file.

### CODE REVIEW

**Comment by Kethku** [comment]
I think this needs rebased on master. Not sure why git is being strange...

File: src/renderer/cursor_renderer/mod.rs [hunk 1]
```diff
@@ -1,23 +1,23 @@
 mod blink;
 mod cursor_vfx;
 
+// use neovide_derive::SettingGroup;
```
Comment (Kethku) [inline src/renderer/cursor_renderer/mod.rs:4]: comment left behind?

File: src/renderer/cursor_renderer/mod.rs [hunk 2]
```diff
@@ -32,54 +32,24 @@ pub struct CursorSettings {
...
-    );
-    register_nvim_setting!(
-        "cursor_vfx_particle_phase",
-        CursorSettings::vfx_particle_phase
-    );
-    register_nvim_setting!(
-        "cursor_vfx_particle_curl",
-        CursorSettings::vfx_particle_curl
-    );
+impl Default for CursorSettings {
```
Comment (Kethku) [inline src/renderer/cursor_renderer/mod.rs:35]: This is WAY cleaner. Seems like it enforces a naming scheme for the resulting option names inside of neovim?

File: src/renderer/cursor_renderer/mod.rs [hunk 3]
```diff
@@ -1,23 +1,23 @@
...
 use blink::*;
 
 const DEFAULT_CELL_PERCENTAGE: f32 = 1.0 / 8.0;
 
 const STANDARD_CORNERS: &[(f32, f32); 4] = &[(-0.5, -0.5), (0.5, -0.5), (0.5, 0.5), (-0.5, 0.5)];
 
-// ----------------------------------------------------------------------------
-
-#[derive(Clone)]
+#[setting_prefix = "cursor"]
```
Comment (Kethku) [inline src/renderer/cursor_renderer/mod.rs:19]: aw I see. Perfect. Setting_prefix solves the issue below. Awesome

**Comment by Kethku** [comment]
Generally this looks awesome!
I had a quick question though. Did this change the names of any existing settings? In a quick scan I don't think it did, but I wanted to make sure.
Other than that, if you rebase (looks like bridge/mod.rs has some unwanted changes) I think I'll merge this. Love the readability and succinctness improvement!

**Comment by Kethku** [comment]
Ah yup I see your note in the pr message. I saw that in another branch as well. I believe if you rebase this branch on top of master the issue should go away.

**Comment by Kethku** [comment]
I think I'm gonna pull this into another branch and rebase it myself

**Comment by Kethku** [comment]
Thanks again so much for the contribution. This is a great improvement in readability and succinctness. Great work

**Comment by tim-harding** [comment]
> 
> 
> Thanks again so much for the contribution. This is a great improvement in readability and succinctness. Great work

Cheers mate, and sorry for the radio silence. I've been meaning to follow up on some things for this project but have been diverted by the start of a new school quarter. Hoping to get back on board at some point.

**Pull request closed without merging**